From 301dd8889dfea9a019f3528428b0d0198e510d1d Mon Sep 17 00:00:00 2001
From: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
Date: Thu, 23 Jan 2025 12:31:59 +0530
Subject: [PATCH] UPSTREAM:PLATFORM:wifi:ath12k:Change firmware stop sequence
 during core deinit

WCSS subsystem CLOCK got disabled in firmware during MODE_OFF this leads to
NOC when host try to access CE register after MODE_OFF during reboot
and rmmod usecases.

Firmware stop done just before the power down avoid accessing
CE register after MODE_OFF. As firmware stop sequence change there might
be interrupt could process if unlock the mutex. Hence register write and
powerdown bring to the single mutex lock block.

Patch-dependency : 711-ath12k-fix-register-access-after-power-down-the-targ.patch

Signed-off-by: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 22 ++++++++++------------
 1 file changed, 10 insertions(+), 12 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index cef3cc6..38a3148 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1191,9 +1191,6 @@ static void ath12k_core_stop(struct ath12k_base *ab)
 
 	ab->ag->num_started--;
 
-	if (!test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags))
-		ath12k_qmi_firmware_stop(ab);
-
 	ath12k_mgmt_rx_reo_deinit_context(ab);
 	ath12k_hif_stop(ab);
 	ath12k_wmi_detach(ab);
@@ -3930,6 +3927,7 @@ static void ath12k_core_hw_group_stop(struct ath12k_hw_group *ag)
 
 	for (i = ag->num_chip - 1; i >= 0; i--) {
 		ab = ag->ab[i];
+		dp = &ab->dp;
 
 		mutex_lock(&ab->core_lock);
 
@@ -3939,15 +3937,6 @@ static void ath12k_core_hw_group_stop(struct ath12k_hw_group *ag)
 		ath12k_dp_umac_reset_deinit(ab);
 		clear_bit(ATH12K_FLAG_CORE_REGISTERED, &ab->dev_flags);
 
-		mutex_unlock(&ab->core_lock);
-	}
-
-	for (i = ag->num_chip - 1; i >= 0; i--) {
-		ab = ag->ab[i];
-		dp = &ab->dp;
-
-		mutex_lock(&ab->core_lock);
-
 		if (!ab->hw_params->reoq_lut_support)
 			break;
 
@@ -3967,10 +3956,15 @@ static void ath12k_core_hw_group_stop(struct ath12k_hw_group *ag)
 				ath12k_hif_write32(ab, address, 0);
 			}
 		}
+
+		if (!test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags))
+			ath12k_qmi_firmware_stop(ab);
+
 		ath12k_hif_power_down(ab);
 
 		mutex_unlock(&ab->core_lock);
 	}
+
 	ath12k_mac_destroy(ag);
 
 	ath12k_core_hw_group_destroy(ag);
@@ -3994,6 +3988,10 @@ void ath12k_core_deinit(struct ath12k_base *ab)
 			ath12k_dp_free(ab);
 			clear_bit(ATH12K_FLAG_CORE_REGISTERED, &ab->dev_flags);
 		}
+
+		if (!test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags))
+			ath12k_qmi_firmware_stop(ab);
+
 		ath12k_hif_power_down(ab);
 		ath12k_core_soc_destroy(ab);
 
-- 
2.34.1

